### ISSUE TITLE
Rules wrongly overrides some Entity functions

### ISSUE THREAD

**Issue by docwilmot** [backdrop-contrib/rules#182]
In a [recent core PR discussion](https://github.com/backdrop/backdrop/pull/4082#discussion_r884191749) we recognized that the `RulesConfigEntity()` class overrides the required `EntityInterface bundle()` method but returns nothing from that. Looking at the class closer, a number of its overriding methods also dont return anything. Furthermore the classes `UcOrderProduct, UcOrder, and UcCartItem` do the same. These should be fixed.

**Comment by bugfolder** [comment]
> Furthermore the classes `UcOrderProduct`, `UcOrder`, and `UcCartItem` do the same. These should be fixed.

They're Ubercart classes, so adding an issue over there. https://github.com/backdrop-contrib/ubercart/issues/339

**Comment by argiepiano** [comment]
After digging a bit, there are a couple of considerations that make this a moot point. 

The entity `rules_config` is a wrapper that allows the controller to access different types of custom rule entity classes that are only used by Rules and are very different from core's (e.g. `RulesReactionRule` or `RulesCondition` etc). They don't extend `Entity`, but are rather their own classes.

So in fact, when these entities are loaded into memory, the class `RulesConfigEntity` is never used to hold the entity. Rather, Rules will instantiate these entities using the "plugin classes" `RulesReactionRule` or `RulesCondition` etc. 

So, even if we create a patch for `RulesConfigEntity::bundle()` that returns `rule_config` (the entity type), the method `bundle()` will still return NULL, as this method doesn't exist in any of the "plugin classes" I mentioned above.

**Comment by argiepiano** [comment]
PR #183 provided. 

Still, this fix is just for the "looks" as it won't make any difference, as explained above.

**Comment by bugfolder** [comment]
Since `RulesConfigEntity` extends `Entity`, and `Entity::bundle()` [returns the entity type by default](https://docs.backdropcms.org/api/backdrop/core%21modules%21entity%21entity.class.inc/function/Entity%3A%3Abundle/1), would it be simpler to just not override `bundle()`?

**Comment by argiepiano** [comment]
[EDIT] Scratch what I just said.

Yes, it'd be easier to not override it. $this->entityType() is defined within `RulesConfigEntity`


~~`Entity::bundle()` returns $this->entityType(), but $this->entityType() is not defined in Entity.~~

~~The idea is that Entity is an abstract class that is there as an example of what to do.~~

**Comment by bugfolder** [comment]
Yes, but `RulesConfigEntity::entityType()` is defined and returns `'rules_config'`.

**Comment by argiepiano** [comment]
Yes, I edited my message above just as you sent yours 😄

**Comment by bugfolder** [comment]
Scratch noted. 😉

**Comment by argiepiano** [comment]
Just pushed that adjustment

**Comment by bugfolder** [comment]
Question about the implementation of `RulesConfigEntity::label()`. There's no member variable `$label` in RulesConfigEntity, Entity, or EntityInterface. Will `return $this->label;` even work?

**Comment by argiepiano** [comment]
That's a good question. The answer is yes, the property is created dynamically when the entity is loaded, even if it's not declared as a `var` within the entity class.

This magic happens in the Controller, because:
1. buildQuery includes all "fields" (columns) declared in the schema when building the database query
2.  And then this line in the controller creates the entity class with all properties based on the columns from the table:

```php
$query_result->setFetchMode(PDO::FETCH_CLASS, $this->entityInfo['entity class'], array(array(), $this->entityType));`
```

So, while it's good to declare those properties explicitly (like in `Node`), it's not necessary. 

In short, $this->label will work because the rules_config table has a column called `label`.

All that said, because of what I explained in a comment above, this entity `rules_config` is a wrapper, and the class RulesConfigEntity never gets instantiated. So, moot point anyway.

**Comment by argiepiano** [comment]
As an experiment to prove this, you can edit the Node class definition and comment out all the property declarations (e.g. nid, title, etc). Clear caches, and run 
```php
dpm(node_load(1));
```

The resulting entity displayed on the screen still has all those properties as part of it.

**Comment by bugfolder** [comment]
I'll take your word for it! Thanks.

**Comment by docwilmot** [comment]
I'll be argumentative and say that the original core discussion is evidence that it isn't moot, because if we don't have proper entities, at some point some code will want to use entity methods and discover that they don't work. Other example being that the rules config entities are loaded normally, with entity_load, and thus invoke entity hooks. Would be best if contrib could expect proper entities passed in any implementations.

**Comment by argiepiano** [comment]
@docwilmot, agreed. See my PR - it includes the required changes.

**Comment by bugfolder** [comment]
Having now re-read all the discussion and related issues, it seems to me that this is OK to merge the PR (and it's getting time for a bugfix release of the module). Any objections?
